src/graphics/aurora/flashfont.cpp

```diff
@@ -0,0 +1,208 @@
+ *
+ * xoreos is the legal property of its developers, whose names
+ * can be found in the AUTHORS file distributed with this source
+ * distribution.
+ *
+ * xoreos is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 3
+ * of the License, or (at your option) any later version.
+ *
+ * xoreos is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with xoreos. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/** @file
+ *  Class for rendering swf fonts.
+ */
+
+#include "src/common/bitstream.h"
+#include "src/common/encoding.h"
+#include "src/common/rect.h"
+
+#include "src/aurora/gfxfile.h"
+
+#include "src/graphics/aurora/flashfont.h"
+
+#ifdef ENABLE_CAIRO
+
+static cairo_user_data_key_t kUserDataKey;
+
+namespace Graphics {
+
+namespace Aurora {
+
+FlashFont::FlashFont(::Aurora::GFXCharacter::Font font) {
+	_ascent = static_cast<double>(font.fontAscent.get());
+	_descent = static_cast<double>(font.fontDescent.get());
+	_advanceTable = font.advanceTable.get();
+	_kernings = font.kerningCodes;
+
+	for (const auto &glyph : font.glyphs) {
+		_glyphs[glyph.code].glyph = glyph;
+
+		_glyphs[glyph.code].width = 0.0;
+		double x = 0.0;
+		for (const auto &shapeRecord : glyph.shapeRecords) {
+			if (shapeRecord.move.deltaX || shapeRecord.move.deltaY) {
+				x = static_cast<double>(shapeRecord.move.deltaX) / 1024.0;
+			} else if (shapeRecord.straightEdge.deltaX || shapeRecord.straightEdge.deltaY) {
+				x += static_cast<double>(shapeRecord.straightEdge.deltaX) / 1024.0;
+			} else if (shapeRecord.curvedEdge.anchorDeltaX || shapeRecord.curvedEdge.anchorDeltaY) {
+				x += (static_cast<double>(shapeRecord.curvedEdge.anchorDeltaX) + static_cast<double>(shapeRecord.curvedEdge.controlDeltaX)) / 1024.0;
+			}
+
+			_glyphs[glyph.code].width = MAX(x, _glyphs[glyph.code].width);
+		}
+	}
+
+	_font = cairo_user_font_face_create();
+
+	if (cairo_font_face_set_user_data(_font, &kUserDataKey, this, 0) != CAIRO_STATUS_SUCCESS)
+		throw Common::Exception("Unable to set cairo font face user data");
+
+	cairo_user_font_face_set_render_glyph_func(_font, &FlashFont::renderGlyph);
+	cairo_user_font_face_set_text_to_glyphs_func(_font, &FlashFont::textToGlyphs);
+}
+
+FlashFont::~FlashFont() {
+	cairo_font_face_destroy(_font);
+}
+
+cairo_status_t FlashFont::renderGlyph(cairo_scaled_font_t *scaled_font, unsigned long glyph, cairo_t *cr,
+									  cairo_text_extents_t *extents) {
+	FlashFont *font = reinterpret_cast<FlashFont *>(
+			cairo_font_face_get_user_data(
+					cairo_scaled_font_get_font_face(scaled_font),
+					&kUserDataKey
+			)
+	);
+
+	info("%i", static_cast<uint16>(glyph));
```

Comment:
`%i` (which is the same as `%d` for printf) is signed. `glyph` is an unsigned long, so `%ul` would be the correct format expression, and you don't need a `static_cast` there then.